google-cloud-storage/src/main/java/com/google/cloud/storage/BidiUploadState.java

```diff
@@ -0,0 +1,1099 @@
+  }
+
+  boolean offer(@NonNull BidiWriteObjectRequest e) {
+    return unimplemented();
+  }
+
+  void updateStateFromResponse(BidiWriteObjectResponse response) {
+    unimplemented();
+  }
+
+  @NonNull GrpcCallContext enqueueFirstMessageAndGetGrpcCallContext() {
+    return unimplemented();
+  }
+
+  void sendVia(Consumer<BidiWriteObjectRequest> consumer) {
+    unimplemented();
+  }
+
+  void updateFromRedirect(@NonNull BidiWriteObjectRedirectedError redirect) {
+    unimplemented();
+  }
+
+  void terminalError() {
+    unimplemented();
+  }
+
+  void pendingRetry() {
+    unimplemented();
+  }
+
+  void retrying() {
+    unimplemented();
+  }
+
+  @Nullable BidiWriteObjectRequest peekLast() {
+    return unimplemented();
+  }
+
+  boolean isFinalizing() {
+    return unimplemented();
+  }
+
+  ApiFuture<Void> beginReconciliation() {
+    return unimplemented();
+  }
+
+  static AppendableUploadState appendableNew(
+      BidiWriteObjectRequest initial,
+      Supplier<GrpcCallContext> baseCallContext,
+      long maxBytes,
+      SettableApiFuture<BidiWriteObjectResponse> resultFuture,
+      @Nullable Crc32cLengthKnown initialCrc32c) {
+    checkArgument(
+        initial.hasWriteObjectSpec(), "provided initial request did not contain a WriteObjectSpec");
+    WriteObjectSpec spec = initial.getWriteObjectSpec();
+    return new NewAppendableUploadState(
+        initial, spec, baseCallContext, maxBytes, resultFuture, initialCrc32c);
+  }
+
+  static AppendableUploadState appendableTakeover(
+      BidiWriteObjectRequest initial,
+      Supplier<GrpcCallContext> baseCallContext,
+      long maxBytes,
+      SettableApiFuture<BidiWriteObjectResponse> resultFuture,
+      @Nullable Crc32cLengthKnown initialCrc32c) {
+    checkArgument(
+        initial.hasAppendObjectSpec(),
+        "provided initial request did not contain a AppendableObjectSpec");
+    AppendObjectSpec spec = initial.getAppendObjectSpec();
+    return new TakeoverAppendableUploadState(
+        initial, spec, baseCallContext, maxBytes, resultFuture, initialCrc32c);
+  }
+
+  private static ImmutableMap<String, List<String>> makeHeadersMap(
+      Stream<String> xGoogRequestParamsEntries) {
+    return ImmutableMap.of(
+        "x-goog-request-params",
+        ImmutableList.of(
+            xGoogRequestParamsEntries.filter(Objects::nonNull).collect(Collectors.joining("&"))));
+  }
+
+  /**
+   * Create a single BidiWriteObjectRequest consisting of the same semantic meaning as if doing
+   * first then second.
+   *
+   * @throws IllegalArgumentException if both first and second have checksummedData
+   */
+  static BidiWriteObjectRequest concatenate(
+      BidiWriteObjectRequest first, BidiWriteObjectRequest second) {
+    checkArgument(
+        !(first.hasChecksummedData() && second.hasChecksummedData()),
+        "attempting to merge two requests that both specify checksummed_data");
+    BidiWriteObjectRequest.Builder b = first.toBuilder().mergeFrom(second);
```

Comment:
Does this throw away the data from one of the two writes?

Reply:
The `checkArgument` on line 195-197 guarantees that at most one of the messages contains data. By the time we make it to `mergeFrom`, we know that only 0 or 1 of the messages contains any actual data. If there were a bug in the caller of this method, the exception from `checkArgument` would prevent it from silently making a bad mutation.